Makefile

```diff
@@ -0,0 +1,156 @@
+test-race:    ARGS=-race         ## Run tests with race detector
+$(TEST_TARGETS): NAME=$(MAKECMDGOALS:test-%=%)
+$(TEST_TARGETS): test
+check test tests: fmt lint vendor | $(BASE) ; $(info  running $(NAME:%=% )tests...) @ ## Run tests
+	$Q cd $(BASE) && $(GO) test -timeout $(TIMEOUT)s $(ARGS) $(TESTPKGS)
+
+test-xml: fmt lint vendor | $(BASE) $(GO2XUNIT) ; $(info  running $(NAME:%=% )tests...) @ ## Run tests with xUnit output
+	$Q cd $(BASE) && 2>&1 $(GO) test -timeout 20s -v $(TESTPKGS) | tee test/tests.output
+	$(GO2XUNIT) -fail -input test/tests.output -output test/tests.xml
+
+COVERAGE_MODE = atomic
+COVERAGE_PROFILE = $(COVERAGE_DIR)/profile.out
+COVERAGE_XML = $(COVERAGE_DIR)/coverage.xml
+COVERAGE_HTML = $(COVERAGE_DIR)/index.html
+.PHONY: test-coverage test-coverage-tools
+test-coverage-tools: | $(GOCOVMERGE) $(GOCOV) $(GOCOVXML)
+test-coverage: COVERAGE_DIR := $(CURDIR)/test/coverage.$(shell date -u +"%Y-%m-%dT%H:%M:%SZ")
+test-coverage: fmt lint vendor test-coverage-tools | $(BASE) ; $(info  running coverage tests...) @ ## Run coverage tests
+	$Q mkdir -p $(COVERAGE_DIR)/coverage
+	$Q cd $(BASE) && for pkg in $(TESTPKGS); do \
+		$(GO) test \
+			-coverpkg=$$($(GO) list -f '{{ join .Deps "\n" }}' $$pkg | \
+					grep '^$(PACKAGE)/' | grep -v '^$(PACKAGE)/vendor/' | \
+					tr '\n' ',')$$pkg \
+			-covermode=$(COVERAGE_MODE) \
+			-coverprofile="$(COVERAGE_DIR)/coverage/`echo $$pkg | tr "/" "-"`.cover" $$pkg ;\
+	 done
+	$Q $(GOCOVMERGE) $(COVERAGE_DIR)/coverage/*.cover > $(COVERAGE_PROFILE)
+	$Q $(GO) tool cover -html=$(COVERAGE_PROFILE) -o $(COVERAGE_HTML)
+	$Q $(GOCOV) convert $(COVERAGE_PROFILE) | $(GOCOVXML) > $(COVERAGE_XML)
+
+.PHONY: lint
+lint: vendor | $(BASE) $(GOLINT) ; $(info  running golint...) @ ## Run golint
+	$Q cd $(BASE) && ret=0 && for pkg in $(PKGS); do \
+		test -z "$$($(GOLINT) $$pkg | tee /dev/stderr)" || ret=1 ; \
+	 done ; exit $$ret
+
+.PHONY: fmt
+fmt: ; $(info  running gofmt...) @ ## Run gofmt on all source files
+	@ret=0 && for d in $$($(GO) list -f '{{.Dir}}' ./... | grep -v /vendor/); do \
+		$(GOFMT) -l -w $$d/*.go || ret=$$? ; \
+	 done ; exit $$ret
+
+# Dependency management
+
+glide.lock: glide.yaml | $(BASE) ; $(info  updating dependencies...)
+	$Q cd $(BASE) && $(GLIDE) update -v
+	@touch $@
+vendor: glide.lock | $(BASE) ; $(info  retrieving dependencies...)
+	$Q cd $(BASE) && $(GLIDE) --quiet install -v
+	@ln -nsf . vendor/src
+	@touch $@
+
+# Docker image
+# To pass proxy for Docker invoke it as 'make image HTTP_POXY=http://192.168.0.1:8080'
+.PHONY: image
+image: | $(BASE) ; $(info Building Docker image...)
+ifdef HTTP_PROXY
+	@docker build -t $(TAG) -f $(DOCKERFILE)  $(CURDIR) --build-arg http_proxy=$(HTTP_PROXY)
+else
+	@docker build -t $(TAG) -f $(DOCKERFILE)  $(CURDIR) 
+endif
+
+# Misc
+
+.PHONY: clean
+clean: ; $(info  Cleaning...)	@ ## Cleanup everything
+	@rm -rf $(GOPATH)
+	@rm -rf $(BUILDDIR)/$(BINARY_NAME)
+	@rm -rf test/tests.* test/coverage.*
```

Comment:
With $GOBIN dir created for golint build, we'll need to remove the $GOBIN dir when make clean.
for example:  @rm -rf $(GOBIN)
nit: the new update seems add bin/golint dir in the repo which I guess will only be created when make is executed.

Reply:
the `bin/` directory will contain go-tools executables such as golint,  gocovmerge etc. Should we be removing and adding them on every build? The `bin/` directory is not added to the project repository. So, the make will only create them locally as needed and install required go-tools as well if they are not present.

Reply:
You're right, the bin/ dir is not added in the repo, and as it's a tool that be used for every build, there is no need to remove it and re-build everytime.